src/MultiTypeSymEntry.chpl

```diff
@@ -325,6 +295,17 @@ module MultiTypeSymEntry
             return prefix + s + suffix;
         }
     }
+    
+    proc createSymEntry(len: int, type etype) throws {
+      var a = makeDistArray(len, etype);
+      return new shared SymEntry(a);
+    }
+
+    proc createSymEntry(a: [?D] ?etype) throws {
+      var A = makeDistArray(D.size, etype);
+      A = a;
+      return new shared SymEntry(A);
+    }
```

Comment:
Just noting that the reason for replacing the `init` methods is that we can't throw from an initializer today in Chapel, so we need the helper functions to first create the array and then handle the potential error before using that allocation to then create the sym entry.

Even though we are calling the function `makeDistArray`, that won't always actually create a distributed array, this is something that is set by a `param` value in the `SymArrayDmap` code, where we are only creating distributed arrays if we are actually in a distributed environment, but I'm not sure if that's the exact issue you are remembering here.

Reply:
Yeah, almost certainly a perf hit from this. @bmcdonald3 can you check the overhead with SymEntrySpeedTest.chpl (it almost certainly needs to be updated, but general flow should be right)

@pierce314159 note that in order to support better OOM handling there may be some cases like this where we have to give up a little performance. I expect that to mostly be from not being able to steal arrays and having extra default-init's. None of these should be major, but wanted to note there may be some cases.

I think this particular code path was optimized so we could steal the memory from dead arrays. If the array isn't dead though, we'd just make a copy. That's more of a compiler optimization and I'm not sure we can tell in code that the array is dead, so I don't know offhand if there's a clean way to be able to steal arrays if their dead, otherwise use the new interface to create.

Ben -- let's talk about during perf meeting.

Reply:
ah yah that makes sense! Thanks for the explanation, I just had a vague memory of you adding it and wanted to make sure it wasn't dropped unless it need to be.

I think as long as the perf hit isn't terribly substantial, we def prefer the better OOM handling, so I'm good with this change!

Reply:
Here is what I'm seeing for performance:

```
$ ./SymEntrySpeedTest --n=10000000
250 arrs from scratch  : 1.41s
250 syms from scratch  : 1.22s
250 arrs from dom      : 1.09s
250 arrs from sym      : 1.08s
250 arrs from arr      : 1.73s
250 syms from arr      : 1.70s
250 arrs from dead arr : 0.00s
250 syms from dead arr : 0.01s

# with new OOM wrappers
$ ./SymEntrySpeedTestNew --n=10000000
250 arrs from scratch  : 1.37s
250 syms from scratch  : 2.98s
250 arrs from dom      : 1.20s
250 arrs from sym      : 1.11s
250 arrs from arr      : 1.74s
250 syms from arr      : 4.63s
250 arrs from dead arr : 0.00s
250 syms from dead arr : 5.13s
```

Syms from scratch (1.22s -> 2.98s): `var A2 = new SymEntry(n, int);` -> `var A2 = createSymEntry(n, int);`

Syms from arr (1.70s -> 4.63s):  `var A2 = new SymEntry(A);` -> `var A2 = createSymEntry(A);`

Syms from dead arr (0.01s -> 5.13s): `var A2 = new SymEntry(n, int);` -> `var A2 = createSymEntry(n, int);`

Will keep digging to see if I can find where that overhead is coming from on the first 2 (where the last one is the dead arrays problem that Elliot mentioned).

Reply:
awesome thanks for looking into it